### PR TITLE
Fixes a set of critical and high-confidence bugs

### PR DESCRIPTION
**Pull request by Fingerling42**
### What changed

- Prevented automatic retry from resubmitting extrinsics after a network error.
- Updated token transfers to use `Balances.transfer_allow_death`, which exists in the current runtime.
- Fixed `Datalog.get_item()` latest lookup for historical reads and ring-buffer wraparound.
- Fixed zero-based extrinsic lookup in `ChainUtils`.
- Preserved RPC errors when reading account nonce instead of silently falling back to `0`.
- Made `Subscriber` read events from the exact block being processed.
- Fixed Subscriber address filtering and Liability event handling.
- Fixed CLI input handling when stdin has no trailing newline.
- Added validation for IPFS CID helpers, chain hashes, DigitalTwin topic hashes, and Liability crypto types.
- Fixed RWS day-boundary rounding and missing-device storage handling.

### Tests

Adds or updates regression tests for the fixed behavior, including:
- extrinsic retry safety;
- Datalog index and latest edge cases;
- nonce RPC error handling;
- Subscriber event filtering;
- hash and CID validation;
- RWS boundary cases;
- Liability signature crypto type validation.

### Compatibility notes

Most changes preserve the existing public API and make previously broken and error-prone cases explicit.

The main behavioral change is extrinsic retry handling: write operations are no longer automatically resubmitted after an ambiguous network failure, because the node may already have accepted the transaction.

### CODE REVIEW

File: robonomicsinterface/classes/datalog.py [hunk 1]
```diff
@@ -13,6 +13,16 @@ class Datalog(BaseClass):
     Class for datalog chainstate queries and extrinsic executions.
     """
 
+    def _get_window_size(self, block_hash: tp.Optional[str] = None) -> int:
```
Comment (PaTara43) [inline robonomicsinterface/classes/datalog.py:16]: а оно переменное, что ли?

File: robonomicsinterface/classes/common_functions.py [hunk 2]
```diff
@@ -71,7 +82,7 @@ def transfer_tokens(self, target_address: str, tokens: int, nonce: tp.Optional[i
 
         return self._service_functions.extrinsic(
             "Balances",
-            "transfer",
+            "transfer_allow_death",
```
Comment (PaTara43) [inline robonomicsinterface/classes/common_functions.py:110]: pзачем?

**Review by PaTara43 — COMMENTED** [review]
Я не смог осилить прям все тесты,  но понимаю, что к чему
у меня только вопрос про переменное окно и про allow death. Все остальное понятно, error handling и более грамотное использование функций
Новый subscriber не будет сильно грузить? Ты разворачиваешь каждое сообщение и отовсюду достаешь место, где может быть искомый адрес?

File: robonomicsinterface/classes/datalog.py (hunk 1 above)
Comment (Fingerling42) [reply robonomicsinterface/classes/datalog.py:16]: Ну я не хочу хардкодить WindowSize кольцевого буфера (оно сейчас 128, но мало ли там akru захочет что-то поменять), поэтому мы читаем Datalog.WindowSize из рантайма.

File: robonomicsinterface/classes/common_functions.py (hunk 2 above)
Comment (Fingerling42) [reply robonomicsinterface/classes/common_functions.py:110]: > pзачем?

В текущем рантайме `Balances.transfer` больше нет. `Balances.transferAllowDeath` выбрал, потому что он похож по поведению на старый трансфер — позволяет сделать перевод, который может уменьшить баланс ниже existential deposit. Думаешь надо вместо этого взять `transferKeepAlive`?

Comment (PaTara43) [reply robonomicsinterface/classes/common_functions.py:110]: > > pзачем?
> 
> 
> 
> В текущем рантайме `Balances.transfer` больше нет. `Balances.transferAllowDeath` выбрал, потому что он похож по поведению на старый трансфер — позволяет сделать перевод, который может уменьшить баланс ниже existential deposit. Думаешь надо вместо этого взять `transferKeepAlive`? 

Да!!
Либо флагом выбирать allow death, либо два разных метода, но точно дефолтным должен быть keepalive

**Comment by Fingerling42** [comment]
> Новый subscriber не будет сильно грузить? Ты разворачиваешь каждое сообщение и отовсюду достаешь место, где может быть искомый адрес?

Не. Как и раньше, subscriber читает `System.Events` один раз на блок и проходит по событиям. Дополнительная логика применяется только после проверки, что событие среди интересных. В частности раньше фильтр адреса был слишком грубым и предполагал что нужный адрес лежит в одном из первых `attributes`. Из-за чего можно было пропустить событие, например у `NewLiability(index, technics, economics, promisee, promisor)` или у NewReport, где адрес находится внутри вложенной структуры. Так что ищем event по реальной структуре в райнтайме.

По идее не должно быть жесткой дополнительно нагрузки.

File: robonomicsinterface/classes/common_functions.py (hunk 2 above)
Comment (Fingerling42) [reply robonomicsinterface/classes/common_functions.py:110]: Поправил: 0f201ac253a7f979bd9cccde800f76d7754bae60

**Review by PaTara43 — APPROVED** [review]
